driver-core/src/main/com/mongodb/client/model/expressions/Expression.java

```diff
@@ -0,0 +1,115 @@
+ * define the output in such cases (though the output may be defined within the
+ * execution context - the server - where the expression is evaluated). Users of
+ * this API must mitigate any risk of applying an expression to some type where
+ * resulting behaviour is not defined by this API (for example, by checking for
+ * null, by ensuring that field types are correctly specified). Likewise, unless
+ * otherwise specified, this API does not define the order of evaluation for all
+ * arguments, and whether all arguments to some expression will be evaluated.
+ *
+ * @see Expressions
+ */
+@Evolving
+public interface Expression {
+
+    /**
+     * Returns logical true if the value of this expression is equal to the
+     * value of the other expression. Otherwise, false.
+     *
+     * @param eq the other expression
+     * @return true if equal, false if not equal
+     */
+    BooleanExpression eq(Expression eq);
+
+    /**
+     * Returns logical true if the value of this expression is not equal to the
+     * value of the other expression. Otherwise, false.
+     *
+     * @param ne the other expression
+     * @return true if equal, false otherwise
+     */
+    BooleanExpression ne(Expression ne);
+
+    /**
+     * Returns logical true if the value of this expression is greater than the
+     * value of the other expression. Otherwise, false.
+     *
+     * @param gt the other expression
+     * @return true if greater than, false otherwise
+     */
+    BooleanExpression gt(Expression gt);
+
+    /**
+     * Returns logical true if the value of this expression is greater than or
+     * equal to the value of the other expression. Otherwise, false.
+     *
+     * @param gte the other expression
+     * @return true if greater than or equal to, false otherwise
+     */
+    BooleanExpression gte(Expression gte);
+
+    /**
+     * Returns logical true if the value of this expression is less than the
+     * value of the other expression. Otherwise, false.
+     *
+     * @param lt the other expression
+     * @return true if less than, false otherwise
+     */
+    BooleanExpression lt(Expression lt);
+
+    /**
+     * Returns logical true if the value of this expression is less than or
+     * equal to the value of the other expression. Otherwise, false.
+     *
+     * @param lte the other expression
+     * @return true if less than or equal to, false otherwise
+     */
+    BooleanExpression lte(Expression lte);
+
+    /**
+     * also checks for nulls
+     * @param other
+     * @return
+     */
+    BooleanExpression isBooleanOr(BooleanExpression other);
+    NumberExpression isNumberOr(NumberExpression other);
+    IntegerExpression isIntegerOr(IntegerExpression other);
+    StringExpression isStringOr(StringExpression other);
+    DateExpression isDateOr(DateExpression other);
+    <T extends Expression> ArrayExpression<T> isArrayOr(ArrayExpression<? extends T> other);
```

Comment:
Let's add something like `@MqlUnchecked` (the draft PR: https://github.com/mongodb/mongo-java-driver/pull/1068) to simplify documenting what is not checked by the API, and to make it more noticeable to users. Usage examples:

```java
ArrayExpression {
    @MqlUnchecked(NON_EMPTY)
    T first();
}
```

```java
Expression {
    <T extends Expression> ArrayExpression<@MqlUnchecked T> isArrayOr(ArrayExpression<? extends T> other);
}
```

```java
DocumentExpression {
    @MqlUnchecked(TYPE)
    NumberExpression getNumber(String fieldName);
}
```

Reply:
included as of https://github.com/mongodb/mongo-java-driver/pull/1070